src/dragonball/src/dbs_pci/src/root_device.rs

```diff
@@ -0,0 +1,411 @@
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            // Safe to unwrap because no legal to generate poisoned RwLock.
+            let state = self.state.read().unwrap();
+            let (b, d, f, o) = parse_mmio_address(offset);
+            if let Some(bus) = state.buses.get(&b) {
+                return bus.write_config(d, f, o | (offset as u32 & 0x3), data);
+            }
+        }
+
+        debug!(
+            "Invalid PCI configuration mmio write ({:x}, {})",
+            base.raw_value(),
+            len
+        );
+    }
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn pio_read(&self, base: PioAddress, offset: PioAddress, data: &mut [u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        // Only allow naturally aligned Dword, Word and Byte access.
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            match offset {
+                // Configuration address register
+                0..=3 => {
+                    // Do not expect poisoned lock here.
+                    let io_addr = self.state.read().unwrap().io_addr >> (offset << 3);
+                    if len == 4 {
+                        NativeEndian::write_u32(data, io_addr);
+                    } else if len == 2 {
+                        NativeEndian::write_u16(data, io_addr as u16);
+                    } else {
+                        data[0] = io_addr as u8;
+                    }
+                    debug!("=>read offset {}, and io_addr: 0x{:x}", offset, io_addr);
+                    return;
+                }
+                // Configuration data register
+                4..=7 => {
+                    // Do not expect poisoned lock here.
+                    let state = self.state.read().unwrap();
+                    if state.io_addr & 0x8000_0000 != 0 {
+                        let (b, d, f, o) = parse_ioport_address(state.io_addr);
+                        if let Some(bus) = state.buses.get(&b) {
+                            return bus.read_config(d, f, o | ((offset as u32) & 0x3), data);
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        debug!(
+            "Invalid PCI configuration ioport read ({:x}, {})",
+            base.raw_value(),
+            len
+        );
+
+        fill_config_data(data);
+    }
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn pio_write(&self, base: PioAddress, offset: PioAddress, data: &[u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+        // Only allow Dword, Word and Byte access.
+        if len != 4 && len != 2 && len != 1 {
+            return;
+        }
+
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
```

Comment:
Do you forget to check for offset here? Addition, there is a duplicated check at the below.

Reply:
yep, I have removed the duplicated check.